msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -95,4 +100,114 @@ public void onError(Exception exception) {
                     }
                 });
     }
+
+    private MsalOAuth2TokenCache getTokenCache(@NonNull final Context appContext,
+                                               @NonNull final IAccountCredentialCache accountCredentialCache) {
+        return new MsalOAuth2TokenCache(
+                appContext,
+                accountCredentialCache,
+                new MicrosoftStsAccountCredentialAdapter());
+    }
+
+    private IAccountCredentialCache getCurrentAccountCredentialCache(final Context appContext){
+        return new SharedPreferencesAccountCredentialCache(
+                new CacheKeyValueDelegate(),
+                new SharedPreferencesFileManager(
+                        appContext,
+                        SINGLE_ACCOUNT_CREDENTIAL_SHARED_PREFERENCES,
+                        new StorageHelper(appContext)
+                )
+        );
+    }
+
+    /**
+     * Get current account that is persisted in shared preference.
+     *
+     * @param appContext
+     * @return a persisted MultiTenantAccount. This could be null.
+     * */
+    @Nullable
+    private MultiTenantAccount getPersistedCurrentAccount(@NonNull final Context appContext) {
+        final String methodName = ":getPersistedCurrentAccount";
+        final IAccountCredentialCache accountCredentialCache = getCurrentAccountCredentialCache(appContext);
+        final MsalOAuth2TokenCache msalOAuth2TokenCache = getTokenCache(appContext, accountCredentialCache);
+
+        List<AccountRecord> accountRecords = accountCredentialCache.getAccounts();
+        if (accountRecords.size() == 0) {
+            return null;
+        } else if (accountRecords.size() > 1) {
+            com.microsoft.identity.common.internal.logging.Logger.verbose(
+                    TAG + methodName,
+                    "Returned accountRecords are larger than 1. " +
+                            "This is unexpected in Single account mode." +
+                            "Clear the cache and return null.");
+
+            accountCredentialCache.clearAll();
+            return null;
+        }
+
+        AccountRecord accountRecord = accountRecords.get(0);
+        List<IdTokenRecord> idTokenRecordList = msalOAuth2TokenCache.getIdTokensForAccountRecord(null, accountRecord);
+
+        List<ICacheRecord> cacheRecordList = new ArrayList<>();
+        for (final IdTokenRecord idTokenRecord : idTokenRecordList) {
```

Comment:
🐛 This is mixing `AccountRecords` from 1 tenant with `IdTokenRecords` from another -- because `AccountRecords` are _tenant specific_, there should be 1 `AccountRecord` + 1 `IdTokenRecord` _per tenant_.

The `realm` field of an `AccessTokenRecord` or `IdTokenRecord` tells you which tenant it belongs to

Some quick examples:
 #### An `IAccount` where the user is signed-in to only their home tenant...
- 1 `ICacheRecord` with 1 `AccountRecord` & 1 `IdTokenRecord` (both for the home tenant) inside

#### An `IAccount` where the user is signed-in to only a guest tenant...
- 1 `ICacheRecord` with 1 `AccountRecord` & 1 `IdTokenRecord` (both for the home tenant) inside

#### An `IAccount` where the user is signed-in to their home tenant and to a guest tenant...
- 2 `ICacheRecords`
    * 1 is for the home tenant, contains 1 `AccountRecord` + 1 `IdTokenRecord`
    * 1 is for the guest tenant, contains 1 `AccountRecord` + 1 `IdTokenRecord`

If a user were to sign-in to multiple guest tenants, then add additional `ICacheRecords` for each authorized tenant, with each containing an additional `AccessTokenRecord` + `IdTokenRecord`